### PR TITLE
[JENKINS-26481] Patch DGM methods with closure parameters

### PR DESCRIPTION
**Pull request by abayer**
[JENKINS-26481](https://issues.jenkins-ci.org/browse/JENKINS-26481)

Subsumes https://github.com/cloudbees/groovy-cps/pull/51 (which itself subsumes https://github.com/cloudbees/groovy-cps/pull/24).

Note that this is currently a work in progress, adding patched methods one by one, but I wanted to get a PR open for watching the changes as we go.

cc @reviewbybees

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -14,6 +14,7 @@
 
   <properties>
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+    <groovy.version>2.4.11</groovy.version>
```
Comment (jglick) [inline pom.xml]: IIUC this will be usable only in 2.61+, right?

File: src/main/java/com/cloudbees/groovy/cps/impl/DGMPatcher.java [hunk 2]
```diff
@@ -252,9 +253,18 @@ private boolean mark(Object o) {
...
                         LOGGER.log(WARNING, "Can't replace members of ManagedLinkedList", item);
-                    // TODO would be possible, if necessary, by calling ManagedLinkedList.Item.remove on *all* elements, then using ManagedLinkedList.add to recreate the entire list
+                        // TODO would be possible, if necessary, by calling ManagedLinkedList.Item.remove on *all* elements, then using ManagedLinkedList.add to recreate the entire list
                     }
                 }
+            } else if (o instanceof ManagedConcurrentLinkedQueue) {
+                for (Iterator itr = ((ManagedConcurrentLinkedQueue) o).iterator(); itr.hasNext(); ) {
+                    Object item = itr.next();
+                    if (patch(item) != item) {
+                        LOGGER.log(WARNING, "Can't replace members of ManagedConcurrentLinkedQueue", item);
```
Comment (jglick) [inline src/main/java/com/cloudbees/groovy/cps/impl/DGMPatcher.java]: I think you meant

```java
LOGGER.log(WARNING, "Can''t replace members of ManagedConcurrentLinkedQueue: {0}", item);
```

or more nicely

```java
LOGGER.log(WARNING, "Cannot replace members of ManagedConcurrentLinkedQueue: {0}", item);
```

or

```java
LOGGER.log(WARNING, "Can’t replace members of ManagedConcurrentLinkedQueue: {0}", item);
```

File: src/main/java/com/cloudbees/groovy/cps/impl/DGMPatcher.java [hunk 3]
```diff
@@ -252,9 +253,18 @@ private boolean mark(Object o) {
...
+                        // TODO would be possible, if necessary, by calling ManagedConcurrentLinkedQueue.Element.remove on *all* elements, then using ManagedLinkedList.add to recreate the entire list
```
Comment (jglick) [inline src/main/java/com/cloudbees/groovy/cps/impl/DGMPatcher.java]: possibly copy-paste error in comment

**Comment by jglick** [comment]
The manual changes to `CpsDefaultGroovyMethods.groovy` are not scalable, I think. There are just too many methods taking a `Closure` and we can expect more to be added in every Groovy release. I would like to try https://github.com/cloudbees/groovy-cps/issues/7 instead.

**Comment by abayer** [comment]
Closing in favor of #56, which now includes tests from here.

**Pull request closed without merging**